src/main/java/com/icthh/xm/gate/security/oauth2/IdpClientRepository.java

```diff
@@ -0,0 +1,96 @@
+package com.icthh.xm.gate.security.oauth2;
+
+import static com.icthh.xm.commons.tenant.TenantContextUtils.getRequiredTenantKeyValue;
+
+import com.icthh.xm.commons.tenant.TenantContextHolder;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import lombok.RequiredArgsConstructor;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
+
+import org.springframework.stereotype.Component;
+
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class IdpClientRepository implements ClientRegistrationRepository {
+
+    private final TenantContextHolder tenantContextHolder;
+
+    private final Map<String, Map<String, ClientRegistration>> clientsHolder = new ConcurrentHashMap<>();
+
+    /**
+     * Register IDP clients for specified tenant
+     *
+     * @param tenantKey     Tenant key
+     * @param registrations IDP clients
+     */
+    public void setRegistrations(String tenantKey, List<ClientRegistration> registrations) {
+        Assert.notEmpty(registrations, "registrations cannot be empty");
+        this.clientsHolder.put(tenantKey, mapRegistrationIdToClientRegistration(registrations));
+
+        List<String> clientsName = registrations.stream()
+            .map(ClientRegistration::getClientName)
+            .collect(Collectors.toList());
+
+        log.info("IDP clients for tenant [{}] registered: {}", tenantKey, clientsName);
+    }
+
+    public void removeTenantClientRegistrations(String tenantKey) {
+        clientsHolder.remove(tenantKey);
+    }
+
+    private static Map<String, ClientRegistration> mapRegistrationIdToClientRegistration(
+        Collection<ClientRegistration> registrations) {
+        return registrations
+            .stream()
+            .peek(registration -> Assert.notNull(registration, "registrations cannot contain null values"))
+            .collect(Collectors.toUnmodifiableMap(ClientRegistration::getRegistrationId, Function.identity()));
+    }
+
+    @Override
+    public ClientRegistration findByRegistrationId(String registrationId) {
```

Comment:
pls add Lep annotation if it possible here

Reply:
@sergeysenja1992 we decided not to add LEP here from a start as we do not see the cases of using it.
But in the future, we may get back to this question and add if need.